core/blockchain_pruner.go

```diff
@@ -0,0 +1,256 @@
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/utils"
+)
+
+const (
+	// pruneBeaconChainBeforeEpoch Keep 16 epoch
+	pruneBeaconChainBeforeEpoch = 16
+	// pruneBeaconChainBlockBefore Keep 16 epoch blocks
+	pruneBeaconChainBlockBefore = 32768 * pruneBeaconChainBeforeEpoch
+	// maxDeleteBlockOnce max delete block once
+	maxDeleteBlockOnce = 10000
+	// maxMemoryUsage max memory usage is 4 MB
+	maxMemoryUsage = 4 * 1024 * 1024
+	// compactProportion compact proportion is thousandth
+	compactProportion = 1000
+)
+
+type blockchainPruner struct {
+	isPruneBeaconChainRunning int32 // isPruneBeaconChainRunning  must be called atomically
+
+	batchWriter ethdb.Batch
+
+	deleteRange              [][2][]byte
+	startTime                time.Time
+	deletedBlockCount        int
+	deletedValidatorSnapshot int
+	skipValidatorSnapshot    int
+}
+
+func newBlockchainPruner() *blockchainPruner {
+	return &blockchainPruner{}
+}
+
+func (bp *blockchainPruner) initPruneBeaconChain(db ethdb.Database) error {
+	if bp.batchWriter == nil {
+		bp.deleteRange = make([][2][]byte, 0)
+		bp.startTime = time.Now()
+		bp.deletedBlockCount = 0
+		bp.deletedValidatorSnapshot = 0
+		bp.skipValidatorSnapshot = 0
+		bp.batchWriter = db.NewBatch()
+		return nil
+	} else {
+		return errors.New("init when batchWriter already init")
+	}
+}
+
+func (bp *blockchainPruner) Delete(key []byte) error {
+	if bp.batchWriter != nil {
+		err := bp.batchWriter.Delete(key)
+		if err != nil {
+			if isUnrecoverableErr(err) {
+				fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+				os.Exit(1)
+			}
+			return err
+		}
+
+		if bp.batchWriter.ValueSize() >= maxMemoryUsage {
+			err = bp.batchWriter.Write()
+			if err != nil {
+				if isUnrecoverableErr(err) {
+					fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+					os.Exit(1)
+				}
+				return err
+			}
+
+			bp.batchWriter.Reset()
+		}
+		return nil
+	} else {
+		return errors.New("delete when batchWriter not init")
+	}
+}
+
+func (bp *blockchainPruner) endPruneBeaconChain() error {
+	if bp.batchWriter != nil {
+		err := bp.batchWriter.Write()
+		if err != nil {
+			if isUnrecoverableErr(err) {
+				fmt.Printf("Unrecoverable error when writing leveldb: %v\nExitting\n", err)
+				os.Exit(1)
+			}
+			return err
+		}
+
+		bp.batchWriter = nil
+		return nil
+	} else {
+		return errors.New("end when batchWriter not init")
+	}
+}
+
+func (bp *blockchainPruner) runWhenNotPruning(cb func()) {
+	// check is start
+	if atomic.CompareAndSwapInt32(&bp.isPruneBeaconChainRunning, 0, 1) {
+		defer atomic.StoreInt32(&bp.isPruneBeaconChainRunning, 0)
+
+		cb()
+	}
+}
+
+func (bp *blockchainPruner) StartPruneBeaconChain(db ethdb.Database, maxBlockNum uint64, maxEpoch *big.Int) {
+	bp.runWhenNotPruning(func() {
+		err := bp.initPruneBeaconChain(db)
+		if err != nil {
+			utils.Logger().Info().Err(err).Msg("pruneBeaconChain init error")
```

Comment:
This is actually not a very good coding practice in harmony code base. I suggest return the error and let caller function to handle/log the errors in lower modules.

Reply:
You are right, edited.